core/src/crds_gossip_push.rs

```diff
@@ -67,10 +70,55 @@ impl CrdsGossipPush {
     pub fn num_pending(&self) -> usize {
         self.push_messages.len()
     }
+
+    pub fn prune_received_cache(
+        &mut self,
+        self_pubkey: &Pubkey,
+        value: VersionedCrdsValue,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> Vec<Pubkey> {
+        let hash = value.value_hash;
+        let origin = value.value.pubkey();
+        let origin_stake = stakes.get(&origin);
+        let self_stake = stakes.get(&self_pubkey);
+        let cache = self.received_cache.get(&hash);
+        if cache.is_none() || origin_stake.is_none() || self_stake.is_none() {
+            return Vec::new();
+        }
+
+        let self_stake = *self_stake.unwrap();
+        let origin_stake = *origin_stake.unwrap();
+        let min_path_stake = self_stake.min(origin_stake);
+        let peer_stake_threshold: u64 =
+            (CRDS_GOSSIP_PRUNE_STAKE_THRESHOLD_PCT * min_path_stake as f64).round() as u64;
+
+        let peers = &cache.unwrap().1;
+        let staked_peers: Vec<(Pubkey, u64)> = peers
+            .iter()
+            .filter_map(|p| stakes.get(p).map(|s| (*p, *s)))
+            .filter(|(_, s)| *s > peer_stake_threshold)
```

Comment:
I was thinking you could keep any number of nodes that add up to the threshold.

Reply:
@aeyakovenko Ok, earlier you suggested randomly prune nodes. Did you mean that it should be done in way that doesn't break the condition that the remaining unpruned nodes still satisfy this condition of adding up to the threshold?

Reply:
> Did you mean that it should be done in way that doesn't break the condition that the remaining unpruned nodes still satisfy this condition of adding up to the threshold?

Yep.  I think the only thing you would need change is after the shuffle remove nodes while the remaining are at least at the threshold.